[IMP] mail, sms, snailmail: adapt mail.notification to OWL

seb-odoo opened the pull request:
task-2089873
task-2171886
task-2153846

seb-odoo commented:
@alexkuhn can you make a first pass on this? This is fully functionally working for the failure (notification group) in the messaging menu.

I still need to do the red envelope in message but this can be done almost completely separately (just using the `notifications` relation on `message`).

seb-odoo commented:
Big diff is due to a lot more tests.

alexkuhn commented:
Do you confirm this PR covers all these tasks?
https://www.odoo.com/web#id=2089873&action=3206&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720
https://www.odoo.com/web#id=2153846&action=3206&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720
https://www.odoo.com/web#id=2171886&action=3206&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720

alexkuhn commented on addons/mail/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:
```diff
@@ -226,6 +226,59 @@ QUnit.test('basic rendering', async function (assert) {
+
+    await this.start({
+        async mockRPC(route, args) {
+            if (args.method === 'channel_fetch_preview') {
+                return [];
+            }
```
don't think you need this, because there's already a mocked `channel_fetch_preview` in `mail/mock_server`

seb-odoo replied:
Correct, wasn't it not the case before? Or maybe I just copy/pasted it from elsewhere.

seb-odoo commented:
Yep, the current code is not fully done yet (issue with popover on message), but the PR will cover all those tasks.

alexkuhn commented on addons/mail/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:
```diff
@@ -226,6 +226,59 @@ QUnit.test('basic rendering', async function (assert) {
+    );
+
+    await afterNextRender(() => {
+        this.env.entities.Notification.insertFromServerData({
+            11: {
+                last_message_date: moment.utc().format("YYYY-MM-DD HH:mm:ss"),
+                message_id: 11,
+                message_type: 'email',
+                model: 'mail.channel',
+                model_name: "Channel",
+                notifications: {
+                    21: {
+                        failure_type: 'SMTP',
+                        notification_id: 21,
+                        notification_status: 'exception',
+                        notification_type: 'email',
+                        partner_id: 41,
+                        partner_name: "Someone",
+                        res_id: 31,
+                    },
+                },
+                res_id: 31,
+            },
+        });
```
ok, though I prefer simulating reception of longpolling notification than calling an internal method like `insertFromServerData()`

seb-odoo replied:
This is arguable. Here it's a component test, not an integration test. The goal of testing a component should be to test that it is rendering correctly depending on existing entity data, and this was the simplest way to create that data.
I think it's best if the test can work even if `MessagingNotificationHandler` is not started.

seb-odoo replied:
At the end I changed it to rely on init messaging data, similar to the other tests related to this.

alexkuhn commented on addons/mail/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:
```diff
@@ -226,6 +226,59 @@ QUnit.test('basic rendering', async function (assert) {
+                        notification_type: 'email',
+                        partner_id: 41,
+                        partner_name: "Someone",
+                        res_id: 31,
```
don't see this key in `mail/mail_notification.py:_notification_format()`

seb-odoo replied:
Good catch, the key is already in message, wrong convert of old code.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+        return this.env.entities.NotificationGroup.get(this.props.notificationGroupLocalId);
+    }
+
+    /**
+     * @returns {string|undefined}
+     */
+    image() {
+        if (this.group.notification_type === 'email') {
+            return '/mail/static/src/img/smiley/mailfailure.jpg';
+        }
+    }
```
makes more sense as a getter in entity (which will become an `attr` field)

seb-odoo replied:
I think it can go either way... This is related to the entity, but it is also a purely visual thing used only in the context of this component.

The only reason why it is in JS in the first place, is because we cannot easily patch XML attribute based on a condition like this. If it was, then it would be in the template, so I think it makes sense to keep it in the component JS?

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+    _onClick(ev) {
+        if (this.group.res_id) {
+            if (this.group.model === 'mail.channel') {
+                this.env.entities.Thread.insert({
+                    id: this.group.res_id,
+                    model: this.group.model,
+                }).open();
```
`insert()` is not yet supported on `master_mail_owl` branch as of now. Need to wait for entity fields branch before doing that.

Reason is that insert just rely on `id`, so if `res_id = 1`/`model = 'some.record'` and there's a channel with Id `1`, the `insert` will update channel... That's fixed with entity fields branch

seb-odoo replied:
So, I just change nothing, and when I rebase on fields later today it will work as expected?

alexkuhn replied:
yes

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.scss:
```diff
@@ -0,0 +1,107 @@
+.o_NotificationGroup {
+    @include o-mail-notification-list-item-layout();
+
+    &:hover .o_NotificationGroup_markAsRead {
+        // TODO SEB also mixin this
+        opacity: 1;
+    }
```
avoid specificity bias, would handle hover in template/OWL and have modifier on `o_NotificationGroup_markAsRead` for this style.

seb-odoo replied:
I just copied code from `o_ThreadPreview_markAsRead`, so out of scope? We can make a pass to clean them all later, I think eventually `ThreadPreview` should become a `NotificationGroup` too, so no more scss mixin for this. (mailbot hack for request will have to be a group too)

alexkuhn replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.scss:
```diff
@@ -0,0 +1,107 @@
+
+    &:hover .o_NotificationGroup_markAsRead {
+        // TODO SEB also mixin this
+        opacity: 1;
```
opacity is style, not layout

seb-odoo replied:
again, copy from thread preview...

alexkuhn replied:
not a reason not to put in correct category here, no?

seb-odoo replied:
I prefer to keep them consistent with each other so that they are easier to merge in task-2258605.
And I don't want to start changing those that I didn't modify in the current PR.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list.js:
```diff
@@ -89,6 +90,18 @@ class NotificationList extends Component {
                 };
             });
+        if (props.filter === 'all') {
+            var notificationGroups = this.env.messaging.notificationGroupManager.groups;
```
no `var`

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list.js:
```diff
@@ -89,6 +90,18 @@ class NotificationList extends Component {
+        if (props.filter === 'all') {
+            var notificationGroups = this.env.messaging.notificationGroupManager.groups;
+            notifications = Object.values(notificationGroups)
+                .sort((group1, group2) => {
+                    return group1.date.isAfter(group2.date) ? -1 : 1;
```
inline

seb-odoo replied:
inline what?

alexkuhn replied:
`.sort((group1, group2) => group1.date.isAfter(group2.date) ? -1 : 1))`

Could be in multi-line... By "inline" I meant arrow function with one-liner notation (so no scope with `{ return ... }`).

seb-odoo replied:
ok for no return, still need multiline if we follow 80 limit

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,629 @@
+                    notification_type: 'email',
+                    partner_id: 41,
+                    partner_name: "Someone",
+                    res_id: 31,
```
again, no such key in `mail/mail_notification.py/_notification_format()` (check whole diff)

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,629 @@
+        "should have 1 group name"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_NotificationGroup_name').innerText,
```
any reason to use `innerText` over `textContent`? Don't think textcontent has some style that we want to take into account, and `innerText` triggers a reflow so prefer `textContent` instead (check whole diff)

seb-odoo replied:
No specific reason, I didn't really pay attention and it worked like this.

I'll have to add some spaces here and there in the expected result with `textContent`, but I'll do it.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,629 @@
+    );
+
+    document.querySelector('.o_NotificationGroup').click();
+    assert.verifySteps(['rpc:channel_fold/open']);
```
not fond of RPC assertion, prefer assertion that chat window is visible

seb-odoo replied:
This is a component test for `NotificationList`, it shouldn't care about the chat window. I don't want to pass `hasChatWindow` to `start` if not necessary.

alexkuhn replied:
IMO check presence of chat window is a much better assertion than relying on server-side synchronization feature of chat windows...

seb-odoo replied:
I think both are kinda bad, as the only thing the component being tested is aware about this click is "open thread". This unit test shouldn't even make a distinction between open in discuss or open in chat window...

But I don't see how it would be possible to test this without patching the entity to assert.step inside thread open.

seb-odoo replied:
Maybe this highlights a bigger problem with thread open (that we already know, for other reasons), and if we want to do such things with it, maybe it should be a function in the env so we can replace it, or maybe it should be a trigger so we can catch it?

alexkuhn replied:
Agree both solutions are not perfect, but just wanted to point out that it's not a reason to not choose the much preferred one.

This is a component test and the aim of click is to open the conversation, be it on chat window, discuss, form view, etc... it doesn't care much... Could have a standardize `data-` that identifies a conversation in use by any component. Ideally not relying on technical stuff (so not local-id) but something understandable by humans and could be on UI but isn't just because it shouldn't be highlighted in the UI. If we take the analogy of a component being a painting, then this `data-` is like a signature.

seb-odoo replied:
Well the preferred one to me is the one with the RPC, or even doing a patch of open (to highlight a design issue). I prefer those than instantiating another component that is not required, if we want to clean test utils start, this is one step in the right direction, not introducing more uses of `hasChatWindow` and other similar config.

In component tests, only `ChatWindowManager` tests should instantiate itself.

alexkuhn replied:
I disagree with preferring assertion on chat window server-sync over assertion on chat window presence...

I think your reasoning comes from not having to instantiate chat window manager in test if we don't have, which is great in theory but in practice if you rely on RPC `channel_fold`, it's worse:

Technically, `start()` with `hasChatWindow` should also disable all logic regarding chat windows, so `channel_fold` shouldn't have been called neither. The reason it does is because we were lazy with this param and just limit with mount of chat window manager (and the fact we plan refactoring `start()` made this improvement kinda pointless in short-term).

I really think the lesser evil of most straightforward approach to make a not-so-bad test is to mount chat window manager and assert presence of chat windows. Not mounting it and asserting on RPC makes this test pass by circumstance, and it's just crying for failure as soon as it can in the future.

That's why I much prefer to consider this as an integration test between notification_group_list and chat_window_manager. In that case, it's better to make the most obvious assertion on a chat window being open, and it's definitely it's presence in DOM over the `channel_fold` RPC.

seb-odoo commented:
Updated PR, should contain:
- the full feature now in /mail and /sms (messaging menu + message (list) indicator)
- rebased on fields
- more fixes in popover
- some fixes in entity
- small python format change

Still need to:
- finish to adapt snailmail message indicator override
- remove remaining parts of the old code
- create/adapt tests for message indicator
- maybe make more use of compute (notif group manager / messaging menu counter) but probably not possible now
- make thread preview also a notif group (probably for later too)

seb-odoo commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.js:
```diff
@@ -0,0 +1,51 @@
+        super(...args);
+        useStore(props => {
+            const notifications = props.notificationLocalIds.map(
+                notificationLocalId => this.env.entities.Notification.get(notificationLocalId)
```
__state

seb-odoo commented:
This should be fully functionally OK now + tested.

Feel free to make a pass.

I'll review myself too.

alexkuhn commented on addons/mail/static/src/messaging/component/message/message.xml:
```diff
@@ -85,6 +85,25 @@
+                            <t t-if="message.notifications.length > 0">
+                                <t t-if="message.failureNotifications.length > 0">
+                                    <span class="o_Message_notificationIcon o-error" t-on-click="_onClickFailure">
+                                        <i name="failure_icon" class="fa fa-envelope"/>
```
camelCase for `name`

seb-odoo replied:
ok (for all of them, I won't say ok every time)

alexkuhn commented on addons/mail/static/src/messaging/component/message/message.xml:
```diff
@@ -85,6 +85,25 @@
                             </t>
+                            <t t-if="message.notifications.length > 0">
+                                <t t-if="message.failureNotifications.length > 0">
+                                    <span class="o_Message_notificationIcon o-error" t-on-click="_onClickFailure">
```
icon is on `<i>`, this is the container, so I would name part `notificationIconContainer`

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/message/message.xml:
```diff
@@ -85,6 +85,25 @@
+                                <t t-else="">
+                                    <Popover>
+                                        <span class="o_Message_notificationIcon">
+                                            <i name="notification_icon" class="fa fa-envelope-o"/>
```
same remark as above

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +217,121 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'email',
+        notifications: [['insert', [{
+            id: 11,
+            notification_status: 'sent',
+            notification_type: 'email',
+            partner: [['insert', { id: 12, name: "Someone" }]],
+        }]]],
```
don't wrap in a list for single insert

seb-odoo replied:
Not sure if I like that magic then (command accepting array or not), but ok to change in the context of this PR.

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +217,121 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon',
+        "should display the notification icon"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-envelope-o',
```
`o_Message_notificationIcon` should be on `<i>`, so that the assertion is on icon an just the icon.

seb-odoo replied:
Ok indeed that felt strange, adding class for the container allows to solve this.

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +217,121 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_NotificationPopover_icon .fa-check',
```
same remark as above

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +217,121 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+            id: 11,
+            notification_status: 'exception',
+            notification_type: 'email',
+        }]]],
```
(again) no wrap in list for single object `insert`

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +217,121 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-envelope',
```
again

alexkuhn commented on addons/mail/static/src/messaging/component/message_list/message_list.js:
```diff
@@ -276,6 +276,13 @@ class MessageList extends Component {
         ) {
             return false;
         }
+        if (
+            prevMessage.notifications.length > 0 ||
+            message.notifications.length > 0
+        ) {
+            // notifications are displayed per message
+            return false;
+        }
```
this doesn't make sense. Just because prev or current message has some notifications, it should not prevent squashing messages in a message list (note that message list is basically thread viewer, not notification_group_list/thread_previews)

seb-odoo replied:
This is needed because the `o_Message_header` is not displayed if messages are squashed, so it would be impossible to see the notifications from the 2nd message.

This is similar for moderation and all other conditions in this method.

Note that even functionally, I feel like emails should never be squashed (so maybe even expand this condition), as they are never sent in quick succession like chat messages, they all deserve their own header and stuff probably.

alexkuhn replied:
Ok I see, let's keep like this then. Would clarify comment on why we don't want squash here, the current one is not clear.

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+    constructor(...args) {
+        super(...args);
+        useStore(props => {
+            const group = this.env.entities.NotificationGroup.get(props.notificationGroupLocalId).__state;
```
no `__state` inspection here, should have ternary condition to handle mutation state concurrency issues

seb-odoo replied:
Do we really want to add those conditions when they are not manifesting? Isn't the goal to get rid of them?

alexkuhn replied:
maybe you don't have this problem on your branch but it sucks if other dev branches have this issue, waste time fixing this issue (for xdu and jgi, it may take some time before they get this is where the issue comes from), while you could have just guarded it in this PR before merge.

seb-odoo replied:
I disagree but this is not important enough to care. I'll just add it to be safe.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+     */
+    _openDocuments() {
+        const device = this.env.messaging.device;
+        if (this.group.notification_type === 'email') {
+            this.env.do_action({
+                name: this.env._t("Mail Failures"),
+                type: 'ir.actions.act_window',
+                view_mode: 'kanban,list,form',
+                views: [[false, 'kanban'], [false, 'list'], [false, 'form']],
+                target: 'current',
+                res_model: this.group.model,
+                domain: [['message_has_error', '=', true]],
+            });
+        }
+        if (!device.isMobile) {
+            this.env.messaging.messagingMenu.close();
+        }
```
all this should be in entities. `messagingMenu.close()` is a reaction to this action, and should be processed a bit like component hints in thread viewer (except it stays in entity world).

seb-odoo replied:
I'm not sure if this comment implies changing more than just moving into entity, but ok to move in entity, you can comment it again after if needed.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+    _onClick(ev) {
+        if (this.group.res_id) {
+            if (this.group.model === 'mail.channel') {
+                this.env.entities.Thread.insert({
+                    id: this.group.res_id,
+                    model: this.group.model,
+                }).open();
```
split in 2 statements (readability)

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,125 @@
+     */
+    _onClickFailureDiscard(ev) {
+        ev.stopPropagation();
+        const device = this.env.messaging.device;
+        if (this.group.notification_type === 'email') {
+            this.env.do_action('mail.mail_resend_cancel_action', {
+                additional_context: {
+                    default_model: this.group.model,
+                    unread_counter: this.group.notifications.length,
+                },
+            });
+        }
+        if (!device.isMobile) {
+            this.env.messaging.messagingMenu.close();
+        }
```
again, makes more sense in entity

seb-odoo replied:
yep, moved all of those

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list.js:
```diff
@@ -93,6 +94,18 @@ class NotificationList extends Component {
+                    group1.date.isAfter(group2.date) ? -1 : 1
+                ).map(notificationGroup => {
+                    return {
+                        notificationGroup,
```
I think it should observe `.__state`, and the getter `notifications` should no longer rely on `storeProps`.

seb-odoo replied:
It should, but the problem is that thread preview is still relying on storeProps, and I don't want the current PR to change more than necessary in that regard.
This is basically task-2258605.

alexkuhn replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,575 @@
+            const NotificationListComponent = components.NotificationList;
+            NotificationListComponent.env = this.env;
+            this.component = new NotificationListComponent(null, { filter });
+            await afterNextRender(async () => {
+                await this.component.mount(this.widget.el);
+            });
```
inline? Or at least "inline" (=non-scoped) arrow function

seb-odoo replied:
I thought that when we did an async function that we didn't care about the return value it should do await inside and not return the result?

alexkuhn replied:
indeed, it prevents some potential programming errors

seb-odoo replied:
So I can ignore your comment here and keep it as it is?

alexkuhn replied:
alright

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,575 @@
+    await this.start({
+        hasChatWindow: true, // needed to assert thread.open
+        async mockRPC(route, args) {
+            if (args.method === 'channel_fold') {
+                return;
+            }
```
I would make a default mockChannelFold RPC in mail/mock_server, so that this is not necessary here

seb-odoo replied:
Yes! I will not argue against you when you actually want to factorize stuff in tests.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.js:
```diff
@@ -0,0 +1,51 @@
+        super(...args);
+        useStore(props => {
+            const notifications = props.notificationLocalIds.map(
+                notificationLocalId => this.env.entities.Notification.get(notificationLocalId).__state
```
dangerous `.__state` if get is `undefined` (mutating concurrency issues)

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.xml:
```diff
@@ -0,0 +1,31 @@
+        <div class="o_NotificationPopover">
+            <t t-foreach="notifications" t-as="notification" t-key="notification.localId">
+                <div>
+                    <span name="notification_status" class="o_NotificationPopover_icon">
```
this looks like `iconContainer`, `icon` should be on `<i>`, and `name` should be the same as the part in BEM class, so `iconContainer` too

seb-odoo replied:
I'll refactor this whole template, and create a new one for snailmail instead of overriding the original one to completely replace it, which didn't make a lot of sense.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.xml:
```diff
@@ -0,0 +1,31 @@
+                            <i class='fa fa-trash-o' title="Canceled" role="img"/>
+                        </t>
+                    </span>
+                    <span name="partner_name" class="o_NotificationPopover_name" t-esc="notification.partner.nameOrDisplayName"/>
```
also `partner_name` =/= `name`, choose one (same for both, camelCase too)

seb-odoo replied:
I don't think we use name= when we have a class to target so I'll just remove it. And yes partnerName is better.

alexkuhn commented on addons/mail/static/src/messaging/component/thread_preview/thread_preview.scss:
```diff
@@ -6,7 +6,7 @@
 
     &:hover .o_ThreadPreview_markAsRead {
-        display: flex;
+        opacity: 1;
```
should be style category, and specificity issue here so hover should be handle in OWL instead of pure css.

seb-odoo replied:
Will be done in task-2258605

alexkuhn commented on addons/mail/static/src/messaging/entity/notification/notification.js:
```diff
@@ -0,0 +1,69 @@
+            if ('id' in data) {
+                data2.id = data.id;
+            }
+            if ('notification_type' in data) {
+                data2.notification_type = data.notification_type;
+            }
+            if ('notification_status' in data) {
+                data2.notification_status = data.notification_status;
+            }
+            if ('failure_type' in data) {
+                data2.failure_type = data.failure_type;
+            }
```
alphabetical order of `data` keys that are processed (ok for exception in partitioning conversions as attributes/relations)

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/entity/notification_group_manager/notification_group_manager.js:
```diff
@@ -0,0 +1,73 @@
+                // we want to open the actual channel in discuss or chat window
+                // and not its kanban/list/form view.
+                const channelId = thread.model === 'mail.channel' ? thread.id : null;
+                const id = `${thread.model}/${channelId}/${notification.notification_type}`;
+                const group = this.env.entities.NotificationGroup.insert({
+                    id,
```
instead of relying on default unicity of entity from `id` and shaping `id` in a way to be uniquely identifiable, I would override `_findFunctionFromData` and make combination of thread model / id / notification type as a way to uniquely identify a notification group. Could also override `_createInstanceLocalId` for easy-debugging with meaningful localId

seb-odoo replied:
Good idea, this was basically old code from before fields when `id` was the only possible way.

seb-odoo replied:
After investigation, this is too hard to do now, at least until we have related dependencies.

The main points:
- group manager should compute everything
- every notification created in the system should be linked to that manager
- it should filter the notifs based on current partner being author of related notification message
- it should manage its list of groups internally based on the filtered notifs above and their fields (at least `notification.message.originThread`: `id` and `model`, + `notification_type`)
- later thread and odoobot request should be introduced into that mix somehow...

alexkuhn commented on addons/mail/static/src/messaging/entity/notification_group_manager/notification_group_manager.js:
```diff
@@ -0,0 +1,73 @@
+                    model_name: thread.model_name,
+                    notification_type: notification.notification_type,
+                });
+                group.update({ notifications: [['link', notification]] });
```
note that this reuses a notification group and just append notification. `computeGroups()` doesn't seem to handle case of a notification group that should unlink some notifications

seb-odoo replied:
It's not a case that happens in practice but I guess it's safer to handle it too.

alexkuhn commented on addons/mail/static/src/messaging/entity/notification_group_manager/notification_group_manager.js:
```diff
@@ -0,0 +1,73 @@
+        // Public
+        //----------------------------------------------------------------------
+
+        computeGroups() {
```
couldn't make this as a computed field? Looks complex, but it would be better than manually calling this

seb-odoo replied:
I think it's too complex at least until we have related dependencies.

alexkuhn commented on addons/mail/static/src/messaging/entity_field.js:
```diff
@@ -428,9 +428,16 @@ class EntityField {
-                    newVal.push(v);
+                const otherValue = otherField.get(otherEntity);
+                if (otherValue) {
+                    if (otherValue instanceof Array) {
+                        // avoid nested array if otherField is x2many too
+                        for (const v of otherValue) {
+                            newVal.push(v);
+                        }
```
nested array is desirable because flatten list may be the same but the inner lists may have change so it should invoke computes. However, relational fields do not support nested fields at the moment, and this would be handy for related on `x2many.x2many` like here. Don't know how hard it is to make support of that though... Could keep flatten in the meantime, but let's but comment as a FIXME because this could have unintended behaviour like I explained here

seb-odoo replied:
task-2261221

alexkuhn commented on addons/mail/static/src/messaging/entity_field.js:
```diff
@@ -428,9 +428,16 @@ class EntityField {
-                    newVal.push(v);
+                const otherValue = otherField.get(otherEntity);
+                if (otherValue) {
+                    if (otherValue instanceof Array) {
```
should check definition of related field, that it's a relational field of type `x2many`. If this is an attribute containing an array, we definitely want to keep nested arrays.

seb-odoo replied:
It is needed for arrays too, because when targeting with a related field a field that is also a related of `attr` through a o2m it must be flattened (but the type of the middle related  is `attr`).

I think we should not support any type of field that would be an array, this is indicating a relation should be made instead. We should also not support object fields as most comparison/equality logic won't work as expected.

If we really want to support it, this condition should become more complex, involving checking recursively if the target field can be a related of a x2many at some point.

seb-odoo replied:
task-2261221

alexkuhn commented on addons/sms/static/src/messaging/component/message/message.xml:
```diff
@@ -0,0 +1,20 @@
+<templates xml:space="preserve">
+
+    <t t-inherit="mail.messaging.component.Message" t-inherit-mode="extension">
+        <xpath expr="//*[@name='failure_icon']" position="replace">
```
Should adapt from change in `name` (it's on container, and camelCase)

alexkuhn commented on addons/sms/static/src/messaging/component/message/message.xml:
```diff
@@ -0,0 +1,20 @@
+            <t t-else="">$0</t>
+        </xpath>
+
+        <xpath expr="//*[@name='notification_icon']" position="replace">
```
same, adapt from suggested changes

alexkuhn commented on addons/sms/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,172 @@
+QUnit.module('Message', {
+    beforeEach() {
+        utilsBeforeEach(this);
+        this.createMessageComponent = async message => {
+            const MessageComponent = components.Message;
+            MessageComponent.env = this.env;
+            this.component = new MessageComponent(null, {
+                messageLocalId: message.localId,
+            });
+            delete MessageComponent.env;
+            await this.component.mount(this.widget.el);
+        };
+        this.start = async params => {
+            let { env, widget } = await utilsStart(Object.assign({}, params, {
+                data: this.data,
+            }));
+            this.env = env;
+            this.widget = widget;
+        };
```
empty newline before/after function definitions

seb-odoo replied:
I'm not sure what do you mean, but all other tests have the same format as this one in that regard?

seb-odoo replied:
This is the only comment left where I don't know what to do. Please check/specify and you can review the rest again.

seb-odoo replied:
done

alexkuhn commented on addons/sms/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,172 @@
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'sms',
+        notifications: [['insert', [{
+            id: 11,
+            notification_status: 'sent',
+            notification_type: 'sms',
+            partner: [['insert', { id: 12, name: "Someone" }]],
+        }]]],
```
unnecessary array for single element with `insert` command

alexkuhn commented on addons/sms/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,172 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-mobile',
```
same comment as before: `<i>` should contain `notificationIcon` so that `fa-*` assertion is on icon itself.

alexkuhn commented on addons/sms/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,172 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_NotificationPopover_icon .fa-check',
```
again

alexkuhn commented on addons/sms/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,172 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-mobile',
```
again

alexkuhn commented on addons/sms/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,317 @@
+            const NotificationListComponent = components.NotificationList;
+            NotificationListComponent.env = this.env;
+            this.component = new NotificationListComponent(null, { filter });
+            await afterNextRender(async () => {
+                await this.component.mount(this.widget.el);
+            });
```
non-scoped/inline arrow function

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message.js:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * @private
+     */
+    async _doActionMissingFields() {
+        const letterIds = await this.env.rpc({
```
avoid async behaviour in component, this is a sign this should be on entity

seb-odoo replied:
This one feels crap even when moved to entity, but I guess that's a first step.

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message.xml:
```diff
@@ -0,0 +1,26 @@
+<templates xml:space="preserve">
+
+    <t t-inherit="mail.messaging.component.Message" t-inherit-mode="extension">
+        <xpath expr="//*[@name='failure_icon']" position="replace">
```
`failureIconContainer` as name

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message.xml:
```diff
@@ -0,0 +1,26 @@
+            <t t-else="">$0</t>
+        </xpath>
+
+        <xpath expr="//*[@name='notification_icon']" position="replace">
```
`notificationIconContainer` as name

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+QUnit.module('Message', {
+    beforeEach() {
+        utilsBeforeEach(this);
+        this.createMessageComponent = async message => {
+            const MessageComponent = components.Message;
+            MessageComponent.env = this.env;
+            this.component = new MessageComponent(null, {
+                messageLocalId: message.localId,
+            });
+            delete MessageComponent.env;
+            await this.component.mount(this.widget.el);
+        };
+        this.start = async params => {
+            let { env, widget } = await utilsStart(Object.assign({}, params, {
+                data: this.data,
+            }));
+            this.env = env;
+            this.widget = widget;
+        };
```
empty line before/after function definitions

seb-odoo replied:
done

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'snailmail',
+        notifications: [['insert', [{
+            id: 11,
+            notification_status: 'sent',
+            notification_type: 'snail',
+        }]]],
```
unnecessary nested array for single item with `insert`

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_NotificationPopover_icon .fa-check',
```
`<i>` should contain `icon` BEM

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'snailmail',
+        notifications: [['insert', [{
+            id: 11,
+            notification_status: 'canceled',
+            notification_type: 'snail',
+        }]]],
```
unnecessary array on single-item with `insert`

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-paper-plane',
```
`<i>` should contain `notificationIcon` BEM

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_NotificationPopover_icon .fa-trash-o',
```
`<i>` should contain `icon` BEM

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'snailmail',
+        notifications: [['insert', [{
+            id: 11,
+            notification_status: 'ready',
+            notification_type: 'snail',
+        }]]],
```
unnecessary array for single-item `insert` (omitting similar comments from now, but should apply in whole diff)

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-paper-plane',
```
`<i>` should contain `notificationIcon` BEM (omitting similar comments from now, but should apply in whole diff)

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+        "icon should represent snailmail"
+    );
+
+    await afterNextRender(() => {
+        document.querySelector('.o_Message_notificationIcon').click();
+    });
```
non-scoped arrow function

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,578 @@
+    await afterNextRender(() => {
+        document.querySelector('.o_Message_notificationIcon').click();
+    });
+    assert.verifySteps(
+        ['search', 'do_action'],
+        "'search' and 'do_action' should have been called"
+    );
```
at a glance most tests looks a bit too technical in my opinion

seb-odoo replied:
Well I guess this is due to the result of do_action not being testable, can't really do anything about that here.

alexkuhn commented on addons/snailmail/static/src/messaging/component/snailmail_error_dialog/snailmail_error_dialog.js:
```diff
@@ -0,0 +1,106 @@
+        useStore(props => {
+            const message = this.env.entities.Message.get(props.messageLocalId);
+            return {
+                message: message.__state,
+                notification: message.notifications[0].__state,
```
guard, `message` and `message.notifications[0]` may be `undefined`

seb-odoo replied:
ok

alexkuhn commented on addons/snailmail/static/src/messaging/component/snailmail_error_dialog/snailmail_error_dialog.xml:
```diff
@@ -0,0 +1,59 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="snailmail.messaging.component.SnailmailErrorDialog" owl="1">
+        <Dialog contentClass="'o_SnailmailErrorDialog'" title="title" size="'medium'" t-ref="dialog">
```
shouldn't it have top-level `t-if` on `message` at least?

seb-odoo replied:
No idea, moderation ban dialog doesn't have it and I copied it.

seb-odoo replied:
I don't want to waste time on this issue, added `t-if` on `message`.

seb-odoo commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,76 @@
+     * @param {MouseEvent} ev
+     */
+    _onClickFailureDiscard(ev) {
+        // avoid opening documents due to the click handler on the group itself
+        ev.stopPropagation();
```
Shouldn't we actually do the reverse, a bit similar to what we do for global events, that is checking inside `_onClick` that the target is not the failure button?

alexkuhn replied:
don't like `ev.stopPropagation()` whatsoever, so I'm for an alternative solution that lets event propagate.

seb-odoo replied:
Done, this is actually pretty simple and could be done everywhere to get rid of all `stopPropagation` without any functional change.

seb-odoo commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.xml:
```diff
@@ -0,0 +1,29 @@
+        <div class="o_NotificationPopover">
+            <t t-foreach="notifications" t-as="notification" t-key="notification.localId">
+                <div class="o_NotificationPopover_notification">
+                    <t t-if="notification.notification_status === 'sent'">
+                        <i class='o_NotificationPopover_icon fa fa-check' title="Sent" role="img"/>
+                    </t>
+                    <t t-if="notification.notification_status === 'bounce'">
+                        <i class='o_NotificationPopover_icon fa fa-exclamation text-danger' title="Bounced" role="img"/>
+                    </t>
+                    <t t-if="notification.notification_status === 'exception'">
+                        <i class='o_NotificationPopover_icon fa fa-exclamation text-danger' title="Error" role="img"/>
+                    </t>
+                    <t t-if="notification.notification_status === 'ready'">
+                        <i class='o_NotificationPopover_icon fa fa-send-o' title="Ready" role="img"/>
+                    </t>
+                    <t t-if="notification.notification_status === 'canceled'">
+                        <i class='o_NotificationPopover_icon fa fa-trash-o' title="Canceled" role="img"/>
+                    </t>
+                    <span class="o_NotificationPopover_partnerName" t-esc="notification.partner.nameOrDisplayName"/>
```
Due to template repetition, wouldn't we want to make "class" and "title" getters in the JS and define the `<i>` only once?

alexkuhn replied:
agree, makes more sense a single `<i>` with getters.

seb-odoo replied:
done

seb-odoo commented on addons/mail/static/src/messaging/entity/notification_group/notification_group.js:
```diff
@@ -0,0 +1,109 @@
+        /**
+         * Opens the view that allows to cancel all notifications of the group.
+         */
+        openCancelAction() {
```
c before d

seb-odoo commented on addons/sms/static/src/messaging/component/message/message.js:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * @override
+     */
+    _onClickFailure() {
+        if (this.message.message_type === 'sms') {
+            this.env.do_action('sms.sms_resend_action', {
+                additional_context: {
+                    default_mail_message_id: this.message.id,
+                },
+            });
+        } else {
+            this._super(...arguments);
+        }
+    },
```
feels like this could/should be moved into entity too?

alexkuhn replied:
Yes, IMO components should just care for rendering and having intercepts of user interactions (e.g. click, keydown, etc.), and the actual handling of intercepts should be in entity.

seb-odoo replied:
done

seb-odoo commented on addons/sms/static/src/messaging/component/message/message.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-inherit="mail.messaging.component.Message" t-inherit-mode="extension">
+        <xpath expr="//*[@name='failureIcon']" position="replace">
+            <t t-if="message.message_type === 'sms'">
+                <i class="o_Message_notificationIcon fa fa-mobile"/> SMS
+            </t>
+            <t t-else="">$0</t>
+        </xpath>
+
+        <xpath expr="//*[@name='notificationIcon']" position="replace">
+            <t t-if="message.message_type === 'sms'">
+                <i class="o_Message_notificationIcon fa fa-mobile"/> SMS
+            </t>
+            <t t-else="">$0</t>
+        </xpath>
```
Would we want to let the message entity return from a compute/getter "fa fa-mobile" depending on its type so that no template override is needed here?

seb-odoo replied:
no

seb-odoo commented on addons/snailmail/static/src/messaging/component/snailmail_error_dialog/snailmail_error_dialog.js:
```diff
@@ -0,0 +1,107 @@
+     */
+    constructor(...args) {
+        super(...args);
+        useStore(props => {
+            const message = this.env.entities.Message.get(props.messageLocalId);
+            const notifications = message ? message.notifications : undefined;
+            return {
+                message: message ? message.__state : undefined,
+                notifications: notifications.map(notification => notification ? notification.__state : undefined),
+                snailmail_credits_url: this.env.messaging.snailmail_credits_url,
+                snailmail_credits_url_trial: this.env.messaging.snailmail_credits_url_trial,
+            };
+        });
```
missing compareDepth, and missing guard on notifications before map (or default it to array)

seb-odoo replied:
done (default to array)

seb-odoo commented on addons/snailmail/static/src/messaging/component/snailmail_notification_popover/snailmail_notification_popover.js:
```diff
@@ -0,0 +1,54 @@
+            const notifications = message ? message.notifications : undefined;
+            return {
+                message: message ? message.__state : undefined,
+                notifications: notifications.map(notification => notification ? notification.__state : undefined),
```
guard on notifications or default to array

seb-odoo commented on addons/snailmail/static/src/messaging/entity/notification_group/notification_group.js:
```diff
@@ -0,0 +1,52 @@
+                additional_context: {
+                    default_model: this.model,
+                    unread_counter: this.notifications.length,
+                }
```
,

alexkuhn commented on addons/mail/static/src/messaging/entity/message/message.js:
```diff
@@ -147,6 +135,9 @@ function MessageFactory({ Entity }) {
                 if ('record_name' in data && data.record_name) {
                     originThreadData.name = data.record_name;
                 }
+                if ('model_name' in data && data.model_name) {
```
not really fond of `message_format` using `model` to actually mean `res_model`, so here with new entry I would definitely name it `res_model_name`

seb-odoo replied:
I don't understand the reasoning here, if the field is `model` it seems inconsistent to have it's `model_name` counterpart be prefixed.

alexkuhn replied:
I don't know how easy/hard it is to rename `model` to `res_model` in `message_format`, so in doubt I prefer not to change right now. With `model_name` here, however, since it has been introduced by this PR, I would really clarify its name and prefix with `res_`.

alexkuhn commented on addons/mail/static/src/messaging/entity/messaging_notification_handler/messaging_notification_handler.js:
```diff
@@ -362,9 +363,23 @@ function MessagingNotificationHandlerFactory({ Entity }) {
+                    message.update({ author: [['link', this.messaging.currentPartner]] });
+                }
+            }
+            this.env.messaging.notificationGroupManager.computeGroups();
+            // manually force recompute of counter (after computing the groups)
+            this.env.messaging.messagingMenu.update();
```
`this.env.messaging` => `this.messaging`

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/entity/notification/notification.js:
```diff
@@ -0,0 +1,69 @@
+    Notification.fields = {
+        failure_type: attr(),
+        id: attr(),
+        notification_status: attr(),
+        notification_type: attr(),
```
n after m

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/entity/notification_group/notification_group.js:
```diff
@@ -0,0 +1,109 @@
+                    });
+                }
+            } else {
+                this._openDocuments();
```
it's weird to have same name method that just differs from visibility in their name, but private method actually is a special case of opening documents. I would just dump code of `_openDocuments` here

seb-odoo replied:
Private method is used for override. The "other case" of the public method doesn't need to be overridden.

seb-odoo replied:
Basically the public method should be called `openDocumentOrDocuments`, and only the plural one needs the override. Not sure that's a good name though.

alexkuhn replied:
BTW I think the naming "document" was a mistake, and we should have used "record".
From the look of it, it either opens one record or a list of records (the latter is solely based on `model`).

I think we could simply override `openRecords()` (public method) and the override is almost the same (just add `!this.res_id` in conditional)

alexkuhn commented on addons/mail/static/tests/helpers/mock_server.js:
```diff
@@ -294,6 +294,9 @@ MockServer.include({
         if (args.method === 'channel_fetch_preview') {
             return Promise.resolve(this._mockChannelFetchPreview(args));
         }
+        if (args.method === 'channel_fold') {
+            return;
+        }
```
I would prefer an actual mock that make uses of data['mail.channel'] and changes `state`/`is_minimized`

seb-odoo replied:
Out of scope? I don't need it and other mocks just return dummy promises too.

seb-odoo replied:
and we don't want to actually reimplement the whole server in js...

alexkuhn replied:
It's very annoying that we cannot just rely on server routes in JS tests, and I really don't like adding magic assumptions in mockRPC... So the only option left is copying some code in mock server... Until we can finally just use python code somehow.

Changes in python related to Discuss are relatively rare, and this code is very small, so I think it's pretty acceptable to fully implement them in our tests, so that behaviour of tests are as closely consistent with reality.

(You could say tours exist for that reason, but I really don't like them because they do test too many stuffs at the same time, and when they fail it's a really pain to debug)

alexkuhn commented on addons/mail_bot/static/src/messaging/component/messaging_menu/messaging_menu.js:
```diff
@@ -17,7 +17,7 @@ patch(components.MessagingMenu, 'mail_bot.messaging.component.MessagingMenu', {
      */
     _onOdoobotRequestClicked() {
-        const device = this.env.messaging.device;
+        const device = this.messagingMenu.messaging.device;
```
I prefer `this.env.messaging`

seb-odoo replied:
~~Out of scope? I don't need it and other mocks just return dummy promises too.~~
wrong comment

seb-odoo replied:
I prefer to change it but it just highlights bad code anyway.

All clicks on the messaging menu notif/previews should close the menu, there should not be an override needed for this in `mail_bot`, and also the code should be in entity and not component...

I guess this should be part of clean task-2258605.

alexkuhn commented on addons/snailmail/static/src/messaging/entity/message/message.js:
```diff
@@ -0,0 +1,65 @@
+     * Opens the action about 'snailmail.letter' missing fields.
+     */
+    async openMissingFieldsLetterAction() {
+        const letterIds = await this.env.rpc({
```
missing `this.async()`

seb-odoo replied:
But `this` is not used asynchronously here, the following code doesn't care if the entity is deleted. The action is valid as long as `letter_id` is known.

alexkuhn replied:
if we add more logic in this function in the future, we could have bugs without `this.async()`. I think we should not decide not to add `this.async()` based on non-failure on current code: any async behaviour should be explicitly tagged as `async`, and deletion of entity should always prevent execution of async code on this entity, regardless of whether this is safe or not now.

seb-odoo commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.xml:
```diff
@@ -0,0 +1,15 @@
+    <t t-name="mail.messaging.component.NotificationPopover" owl="1">
+        <div class="o_NotificationPopover">
+            <t t-foreach="notifications" t-as="notification" t-key="notification.localId">
+                <div class="o_NotificationPopover_notification">
+                    <i class="o_NotificationPopover_icon" t-att-class="iconClass" t-att-title="iconTitle" role="img"/>
+                    <span class="o_NotificationPopover_partnerName" t-esc="notification.partner.nameOrDisplayName"/>
+                </div>
```
The 2 inner classes should contain _notification somewhere in their name?

alexkuhn replied:
Don't see the gain of prefixing part names with `_notification`, since this component is about **Notification**Popover

seb-odoo commented on addons/mail/static/src/messaging/entity/notification_group/notification_group.js:
```diff
@@ -0,0 +1,109 @@
+                    domain: [['message_has_error', '=', true]],
+                });
+            }
+            if (!device.isMobile) {
+                this.env.messaging.messagingMenu.close();
+            }
```
I'm actually wondering if this shouldn't always become flux specific? This only makes sense when clicking on a messaging menu element?
If I have the messaging menu open and I'm opening a document from another mean, I don't really want my messaging menu closed?

seb-odoo replied:
Or it's getting closed automatically anyway due to global click handler.

alexkuhn replied:
I think you can remove this code, and let global click handler auto-close it.

seb-odoo commented:
All changes have been done or commented. Can be reviewed again + check unresolved comments.

alexkuhn commented on addons/mail/static/src/messaging/component/message/message.scss:
```diff
@@ -332,6 +332,15 @@
 }
 
+.o_Message_notificationIconContainer {
+    color: $gray-600;
```
didn't we say we prefer `gray('<value>')` over `$gray-<value>`? This variable is just a alias, and even its naming is too similar with original one, so that's a useless use of variable IMO

seb-odoo replied:
I don't remember but ok.

The variable is not an alias though it is the original value.
The function is the alias, but I guess it could be useful in some kind of customization scheme.

alexkuhn replied:
You're right... Now I don't know which one is better... Maybe keep `gray()` for consistency, as you said it could be useful for some customization maybe?

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +219,131 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+        '.o_NotificationPopover_partnerName',
+        "popover should have the partner name"
+    );
+    assert.equal(
```
`strictEqual`

seb-odoo replied:
ok (I'll do all of them, no need to re-comment)

alexkuhn commented on addons/mail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -217,6 +219,131 @@ QUnit.test('moderation: moderated channel with pending moderation message (moder
+    await afterNextRender(() => {
+        document.querySelector('.o_Message_notificationIconContainer').click();
+    });
+    assert.verifySteps(['do_action'], "'do_action' should have been called");
```
I think I would clarify intent of do action in comment of assertion, something like "should have made an action to resend message", just to better understand the importance of `do_action` in this test, and should really help in future to adapt/remove them (like action manager is incorporated in mock environment, or we just stop using do_action and use something else, etc...).

I know assertions in intercepted do_action give these clues, but if test fails due to do_action not being intercepted for some reasons, it requires some effort to find appropriate information to understand why do_action is needed. Tests has some assertions inside intercepted `do_action` that give some clues to purpose of this `do_action`, but it still doesn't give a big picture of intent of this flow IMO, hence the suggestion to clarify in comment of this assertion.

seb-odoo replied:
Yes it's better, I think I was just lazy to write better message since as you said assert inside action were already descriptive.

alexkuhn commented on addons/mail/static/src/messaging/component/message_list/message_list.js:
```diff
@@ -276,6 +276,15 @@ class MessageList extends Component {
+            prevMessage.notifications.length > 0 ||
+            message.notifications.length > 0
+        ) {
+            // When messages are squashed, `o_Message_header` is only visible
+            // for the first message. This would prevent from seeing
+            // notifications from the 2nd message onwards.
```
Basically could be summarized with "Visual about message notifications is restricted to non-squashed messages."

seb-odoo replied:
Ok I'll simplify it, I added more details since you said it wasn't clear originally.

alexkuhn commented on addons/mail/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:
```diff
@@ -226,6 +227,38 @@ QUnit.test('basic rendering', async function (assert) {
     );
 });
 
+QUnit.test('counter', async function (assert) {
```
name of test is too generic. Here it tests that a single isolated failure notification is included in messaging menu (global) counter.

seb-odoo replied:
counter is taking into account failure notification

alexkuhn commented on addons/mail/static/src/messaging/component/notification_group/notification_group.js:
```diff
@@ -0,0 +1,84 @@
+     * @param {MouseEvent} ev
+     */
+    _onClick(ev) {
+        if (ev.target === this._markAsReadRef.el) {
+            // handled in `_onClickFailureDiscard`
+            return;
+        }
```
The intent is good, but the current solution is not satisfying: it assumes ref has no children and is not a component (if it becomes a component, we have to change this LOC, which kinda sucks... even so if this component also has children and/or sub-components, this becomes annoying to check).

seb-odoo replied:
Changed to `contains` instead of equality.

As for `.el` or `.comp` there is nothing we can do since this is how `useRef` works, we already questioned/reported the annoyance of it, but this shouldn't prevent us from using it.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list.js:
```diff
@@ -68,7 +69,7 @@ class NotificationList extends Component {
     _useStoreSelector(props) {
         const threads = this._useStoreSelectorThreads(props);
-        const notifications = threads
+        let notifications = threads
```
it has become quite misleading to name them "notifications" now that there's a `Notification` entity. I was looking fora relational field `notificationGroup` in `Notification`, just realising `storeProps.notifications` is not a list of `Notification` entities...

seb-odoo replied:
Yes, this must be done as part of the clean up task-2258605

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,569 @@
+    document.querySelector('.o_NotificationGroup_markAsRead').click();
+    assert.verifySteps(
+        ['do_action'],
+        "should have called do_action correctly"
```
again would clarify intent of do_action in comment of such assertions (any other similar assertions should have same treatment)

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,569 @@
+    assert.containsOnce(
+        document.body,
+        '.o_ChatWindow',
+        "should have opened thread"
```
_"... channel of notification failure related to 1st group in a chat window"_

seb-odoo replied:
ok for the clarification except I don't want to explicit "chat window" because this test really does not care if it's open in chat window or discuss

alexkuhn replied:
It shouldn't, but at the moment it does rely on that and it's made clear by assertion on `.o_ChatWindow`. I prefer a clear assertion comment that gives purpose of assertion, even if it's to specific at the moment. The main idea is to help in case of failure, and knowing that it fails because there is no chat window of thread is super helpful.

In the meantime of improving this test so that detection of opening of thread is not chat window-dependent, I prefer being super explicit when the assertion relies on that.

alexkuhn commented on addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js:
```diff
@@ -0,0 +1,569 @@
+    );
+
+    await afterNextRender(() => {
+        document.querySelector('.o_NotificationGroup').click();
```
Would clarify click is on 1st group

seb-odoo replied:
Ok, I can use `groups[0]` since I already have the variable

alexkuhn commented on addons/mail/static/src/messaging/entity/notification_group/notification_group.js:
```diff
@@ -0,0 +1,109 @@
+                if (this.model === 'mail.channel') {
+                    const channel = this.env.entities.Thread.insert({
+                        id: this.res_id,
+                        model: this.model,
+                    });
+                    channel.open();
+                } else {
+                    this.env.messaging.openDocument({
+                        id: this.res_id,
+                        model: this.model,
```
would replace `this.model` by `mail.channel`. I know this require 3 changes of LOC if we ever have to change model of channels, but on one hand it's one less variable to care and bare string is much easier to read than a variable. What do you think?

seb-odoo replied:
I think you meant to comment the one inside the `if (this.model === 'mail.channel')` and not the `else`.

I don't really care either way, I guess this would make sense if we had channel entity, we would do "entity channel open" and just give it the id and not the model.

seb-odoo replied:
But it's not one less variable though, it still relies on the use of this.model on the if itself, and if the code is moved outside of that condition it will break/not make sense.

So I guess it is safer to keep it like it is for now?

alexkuhn commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.xml:
```diff
@@ -0,0 +1,15 @@
+        <div class="o_NotificationPopover">
+            <t t-foreach="notifications" t-as="notification" t-key="notification.localId">
+                <div class="o_NotificationPopover_notification">
+                    <i class="o_NotificationPopover_notification_icon" t-att-class="iconClass" t-att-title="iconTitle" role="img"/>
+                    <span class="o_NotificationPopover_notification_partnerName" t-esc="notification.partner.nameOrDisplayName"/>
```
camelCase for BEM part

seb-odoo replied:
ok

alexkuhn commented on addons/mail/static/src/messaging/component/notification_popover/notification_popover.scss:
```diff
@@ -0,0 +1,7 @@
+// Layout
+// -----------------------------------------------------------------------------
+
+.o_NotificationPopover_notification_icon {
```
camelCase

seb-odoo replied:
ok

alexkuhn commented on addons/mail/wizard/mail_resend_cancel.py:
```diff
@@ -24,7 +24,7 @@ def cancel_resend_action(self):
                                 JOIN mail_message mes
                                     ON notif.mail_message_id = mes.id
-                                WHERE notif.notification_status IN ('bounce', 'exception')
+                                WHERE notif.notification_type = 'email' AND notif.notification_status IN ('bounce', 'exception')
```
looks like stable/master fix

seb-odoo replied:
Fix was done in stable, diff here will disappear at next rebase.

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message.js:
```diff
@@ -0,0 +1,87 @@
+            switch (this.message.notifications[0].failure_type) {
+                case 'sn_credit':
+                    // URL only used in this component, not received at init
+                    this.env.messaging.fetchSnailmailCreditsUrl();
+                    this.snailmailState.hasDialog = true;
```
In theory, if snailmail credit url could change, then we could see an old url in dialog before RPC is resolved. This should encourage either have hasDialog in entity (wait for RPC before display), or manage a "please wait" until valid URL is visible.

seb-odoo replied:
I can just unset the existing value before calling the RPC if we consider the old value unreliable.

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,623 @@
+                );
+                assert.strictEqual(
+                    ev.data.options.additional_context.letter_id,
+                    22,
+                    "action should have correct letter id"
+                );
+            },
+        },
+        async mockRPC(route, args) {
+            if (args.model === 'snailmail.letter' && args.method === 'search') {
+                assert.step('search');
+                return [22];
```
`22` looks like a magic value (just take it for true, not that meaningful of what it means), and then assertion on that magic value. Just pointing another reason to not liking manually mocked RPC.

seb-odoo replied:
I'll add a comment that this id is related to the one passed to the action. But yes nothing else to do here...

alexkuhn commented on addons/snailmail/static/src/messaging/component/message/message_tests.js:
```diff
@@ -0,0 +1,623 @@
+    });
+    assert.verifySteps(
+        ['search', 'do_action'],
+        "'search' and 'do_action' should have been called"
```
meaning of both of them? I guess search for related letter id of message (hard-coded) and do_action should prompt something to fix the issue?

seb-odoo replied:
"the id of the letter related to the message should be returned by a search and an action should be done to display the missing fields dialog"

alexkuhn commented on addons/snailmail/static/src/messaging/component/snailmail_error_dialog/snailmail_error_dialog.xml:
```diff
@@ -0,0 +1,61 @@
+
+                <t t-set="buttons">
+                    <t t-if="hasCreditsError">
+                        <button class="o-resend-letter btn btn-primary" t-on-click="_onClickResendLetter">Re-send letter</button>
+                    </t>
+                    <button class="o-cancel-letter btn"
```
why not structural classnames like `o_SnailmailErrorDialog_resendLetterButton`? Seems more appropriate, and should avoid mistakenly sharing style

seb-odoo replied:
It's just used as a selector in tests for now, but looks like a good idea to rename.

alexkuhn commented on addons/snailmail/static/src/messaging/component/snailmail_error_dialog/snailmail_error_dialog.xml:
```diff
@@ -0,0 +1,61 @@
+                    </p>
+                </t>
+                <t t-elif="notification.failure_type === 'sn_error'">
+                    <p class="o_SnailmailErrorDialog_content_error">
```
camelCase (check most part name of this file)

seb-odoo replied:
ok

alexkuhn commented:
popover on failure is not displayed on mouseover, contrary to master (visual from master, any type of failure):
![master-popover-failure](https://user-images.githubusercontent.com/6569390/82685662-d8667280-9c54-11ea-970b-0feff985cecb.png)

alexkuhn commented:
green (except migration, but that was expected)
